scanpipe/cyclonedx/__init__.py

```diff
@@ -0,0 +1,181 @@
+
+
+def get_components(bom: Bom_1_4):
+    """
+    Return list of components from CycloneDx BOM.
+    """
+    return recursive_component_collector(bom.components, [])
+
+
+def bom_attributes_to_dict(cyclonedx_attributes):
+    """
+    Return list dict from a list of CycloneDx attributes.
+    """
+    if not cyclonedx_attributes:
+        return {}
+
+    return [
+        json.loads(attribute.json(exclude_unset=True, by_alias=True))
+        for attribute in cyclonedx_attributes
+    ]
+
+
+def recursive_component_collector(root_component_list, collected):
+    """
+    Return list of components including the nested components.
+    """
+    if not root_component_list:
+        return
+
+    for component in root_component_list:
+        extra_data = (
+            bom_attributes_to_dict(component.components)
+            if component.components is not None
+            else {}
+        )
+        collected.append({"cdx_package": component, "nested_components": extra_data})
+        recursive_component_collector(component.components, collected)
+    return collected
+
+
+def resolve_license(license):
+    """
+    Return license expression/id/name from license item.
+    """
+    if "expression" in license:
+        return license["expression"]
+    elif "id" in license["license"]:
+        return license["license"]["id"]
+    else:
+        return license["license"]["name"]
+
+
+def get_declared_licenses(licenses):
+    """
+    Return resolved license from list of LicenseChoice.
+    """
+    if not licenses:
+        return ""
+
+    return "\n".join(
+        [resolve_license(license) for license in bom_attributes_to_dict(licenses)]
+    )
+
+
+def get_checksums(component: Component):
+    """
+    Return dict of all the checksums from a component.
+    """
+    if not component.hashes:
+        return {}
+
+    algorithm_map_cdx_scio = {
+        "MD5": "md5",
+        "SHA-1": "sha1",
+        "SHA-256": "sha256",
+        "SHA-512": "sha512",
+    }
+    return {
+        algorithm_map_cdx_scio[algo_hash.alg.value]: algo_hash.content.__root__
+        for algo_hash in component.hashes
+        if algo_hash.alg.value in algorithm_map_cdx_scio
+    }
+
+
+def get_external_refrences(external_references):
+    """
+    Return dict of refrence urls from list of `externalRefrences`.
+    """
+    if not external_references:
+        return {}
+
+    refrences = {
+        "vcs": [],
+        "issue-tracker": [],
+        "website": [],
+        "advisories": [],
+        "bom": [],
+        "mailing-list": [],
+        "social": [],
+        "chat": [],
+        "documentation": [],
+        "support": [],
+        "distribution": [],
+        "license": [],
+        "build-meta": [],
+        "build-system": [],
+        "release-notes": [],
+        "other": [],
+    }
+    for ref in external_references:
+        refrences[ref.type.value].append(ref.url)
+
+    return {key: value for key, value in refrences.items() if value}
```

Comment:
I'm not sure to understand your logic here, why build a dict of empty lists to then remove those in the `return`?
You may want to use a defaultdict instead.